artifacts.go

```diff
@@ -301,207 +292,194 @@ func validateArtifact(c *cli.Context) error {
-				int64(len(signed)), bytes.NewBuffer(signed))); err != nil {
-		return err
+	if err := checkIfValid(c.Args().First(), c.String("key")); err != nil {
+		return cli.NewExitError(err.Error(), 1)
 	}
+
+	fmt.Println("Artifact file '" + c.Args().First() + "' validated successfully")
 	return nil
 }
 
-func processHeader(r *areader.Reader, w *awriter.Writer,
-	key []byte, force bool) error {
-	// simple list with the header element name and the flag uses as
-	// indicator if element is optional or required
-	artifactHeaderElems := []struct {
-		name     string
-		required bool
-	}{
-		{"manifest", true},
-		{"manifest.sig", false},
-		{"header.", true},
-	}
-
-	getNext := true
-	var raw *artifact.Raw = nil
-	var err error = nil
-
-	// read header elements first
-	for _, elem := range artifactHeaderElems {
-		// get element form the artifact
-		if getNext {
-			raw, err = r.ReadRaw()
-			if err != nil {
-				return err
-			}
-		}
-		getNext = true
-
-		// check if we are not having element out of order
-		if elem.required && !strings.HasPrefix(raw.Name, elem.name) {
-			return errors.Errorf("Invalid artifact, should contain '%s' "+
-				"file , but contains '%s'", elem.name, raw.Name)
-		} else if !elem.required && !strings.HasPrefix(raw.Name, elem.name) {
-			// we have missing optional element; move on to the next one
-			getNext = false
-			continue
-		}
+func signExisting(c *cli.Context) error {
+	if c.NArg() == 0 {
+		return cli.NewExitError("Nothing specified, nothing signed. \nMaybe you wanted"+
+			" to say 'artifacts sign <pathspec>'?", 1)
+	}
 
-		// check if we are having "manifest" file, which we need to sign
-		if raw.Name == "manifest" {
-			if err = createSignature(raw, w, key); err != nil {
-				return err
-			}
-			continue
-
-		} else if raw.Name == "manifest.sig" && !force {
-			// we are re-signing the artifact; return error by default
-			return errors.New("Trying to sign already signed artifact; " +
-				"please use force option")
-		} else if raw.Name == "manifest.sig" && force {
-			// just continue here as new signature is already part of tmp artifact
-			continue
-		}
+	if len(c.String("key")) == 0 {
+		return cli.NewExitError("Missing signing key; "+
+			"please use `-k` parameter for providing one", 1)
+	}
 
-		if err = w.WriteRaw(raw); err != nil {
-			return errors.Wrap(err, "Can not write artifact")
-		}
+	privateKey, err := getKey(c.String("key"))
+	if err != nil {
+		return cli.NewExitError("Can not use signing key provided: "+err.Error(), 1)
 	}
-	return nil
-}
 
-func signArtifact(r *areader.Reader, w *awriter.Writer,
-	rawVer *artifact.Raw, key []byte, force bool) error {
-	// first we need to store version in new artifact we are trying to sign
-	if err := w.WriteRaw(rawVer); err != nil {
-		return err
+	tFile, err := ioutil.TempFile("", "mender-artifact")
+	if err != nil {
+		return errors.Wrap(err,
+			"Can not create temporary file for storing artifact")
 	}
+	defer os.Remove(tFile.Name())
```

Comment:
Actually I can do defer. It will simply fail if the file is already closed, but I don't care at that point. The important thing is we are not leaking resources.